source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,209 @@
+#include "common/http/filter/gzip_filter.h"
+
+#include <regex>
+
+#include "common/common/macros.h"
+
+namespace Envoy {
+namespace Http {
+
+namespace {
+// Default zlib memory level.
+const uint64_t DEFAULT_MEMORY_LEVEL{5};
+
+// Default and maximum compression window size.
+const uint64_t DEFAULT_WINDOW_BITS{12};
+
+// Minimum length of an upstream response that allows compression.
+const uint64_t MINIMUM_CONTENT_LENGTH{30};
+
+// When summed to window bits, this sets a gzip header and trailer around the compressed data.
+const uint64_t GZIP_HEADER_VALUE{16};
+
+// Default content types will be used if any is provided by the user.
+const std::unordered_set<std::string> DEFAULT_CONTENT_TYPES{
+    "text/html",        "text/plain",    "text/css", "application/javascript",
+    "application/json", "image/svg+xml", "text/xml", "application/xhtml+xml"};
+
+// Regex used to validate a gzip Accept-Encoding header.
+const std::regex& acceptEncodingRegex() {
+  CONSTRUCT_ON_FIRST_USE(std::regex, "(?!.*gzip;\\s*q=0(,|$))(?=(.*gzip)|(^\\*$))",
```

Comment:
I have a hard to time reading this regex.  Can you (a) comment on what this is intended to accept & reject and (b) add unit tests for it?

I think it's probably easier to split on commas, and see if any of the whitespace-stripped tokens is "gzip".

Reply:
Having a hard to time reading regexes is a good thing :) 
I'm not happy with this solution either. I ran some benchmarks and it seems that optimize barely makes any difference. I will try to replace regex with something using the string_view that you showed me earlier.

Reply:
Curious: did you use a microbenchmarking infrastructure?  Ironically the one I've been using in open-source is from RE2: https://github.com/google/re2/blob/master/util/benchmark.h

Example use: https://github.com/pagespeed/mod_pagespeed/blob/ce1b29534f93763fb38b097567caf719b8277bda/pagespeed/kernel/html/html_parse_speed_test.cc#L21

Reply:
example use of string-splitting:

  std::vector<absl::string_view> tokens = absl::StrSplit(value, ',', absl::SkipEmpty());
then you can loop over the tokens, strip the leading & trailing whitespace, and do lookups without copying any bytes of any of the strings (assuming you use a map of string_views which does require you to have some backing store for the string data in the map).

Reply:
No, my benchmark is much simpler :)

I used this online tool (http://quick-bench.com) that also uses google open-source (https://github.com/google/benchmark).

http://quick-bench.com/7HcNBN2GJXdeKUnSyck1WQ18Ze8 

 absl::StrSplit(value, ',', absl::SkipEmpty()) looks neat!

Reply:
Oh that's really the same thing I was thinking of, just a nicely exported version of it.  Very nice that there's a way to trigger it online.  Will check that out.  Thanks!